### PR TITLE
Stabilize macros in some more positions

### LINKED ISSUE #49476: Tracking issue for `macros_in_extern` feature
This is a tracking issue for the `macros_in_extern`. This feature flag enables the usage of macro invocations in `extern {}` blocks:

```rust
macro_rules! emit_foreign_item(
    () => (fn read(fd: int, buf: *mut u8, size: usize) -> int;)
);

extern {
    emit_foreign_item!();
}
```

Declarative maco, proc-macro and proc-macro-attribute invocations are allowed as well. Macro defintions, however, are not.

This feature is unstable because it was recently implemented in #49350 without an RFC. It is considered a discrepancy in macro support and not a new feature. It should be completely backwards-compatible.

### PR DESCRIPTION
**Pull request by petrochenkov**
- Fn-like macros and attribute macros in `extern` blocks
- Fn-like procedural macros in type positions
- ~Attribute macros on inline modules~ (moved to https://github.com/rust-lang/rust/pull/64273)

Stabilization report: https://github.com/rust-lang/rust/pull/63931#issuecomment-526362396.

Closes https://github.com/rust-lang/rust/issues/49476
cc https://github.com/rust-lang/rust/issues/54727

### CODE REVIEW

File: src/libsyntax/ext/expand.rs [hunk 1]
```diff
@@ -657,7 +650,7 @@ impl<'a, 'b> MacroExpander<'a, 'b> {
             AstFragmentKind::Expr => "expressions",
             AstFragmentKind::OptExpr => "expressions",
             AstFragmentKind::Pat => "patterns",
-            AstFragmentKind::Ty => "types",
+            AstFragmentKind::Ty => return,
```
Comment (Centril) [inline src/libsyntax/ext/expand.rs]: Cleanup: can we merge the `=> return` arms with an or-pattern?

File: src/libsyntax/ext/expand.rs [hunk 2]
```diff
@@ -595,19 +587,20 @@ impl<'a, 'b> MacroExpander<'a, 'b> {
...
             Annotatable::Item(ref item) => {
                 match item.node {
-                    ItemKind::Mod(_) if self.cx.ecfg.proc_macro_hygiene() => return,
-                    ItemKind::Mod(_) => ("modules", sym::proc_macro_hygiene),
+                    ItemKind::Mod(ref m) if !m.inline =>
+                        ("non-inline modules", sym::proc_macro_hygiene),
                     _ => return,
                 }
             }
             Annotatable::TraitItem(_) => return,
```
Comment (Centril) [inline src/libsyntax/ext/expand.rs]: Same here re. or-patterns.

**Comment by Centril** [comment]
> * Fn-like macros and attribute macros in `extern` blocks
> * Fn-like procedural macros in type positions
> * Attribute macros on inline modules

From a language POV these all seem orthogonal and it seems like that's true from the POV of the compiler code as well. As such, I would like us to consider these separately in 3 different stabilization PRs with associated reports & FCPs.

> * Fn-like procedural macros in type positions

In the associated report, it would be good to elaborate re. hygiene, how this may e.g. interact with expressions in type contexts, and give a recap (because issues re. hygiene isn't the best known subject...) of the issues wit proc macros expanding to expressions and why it isn't an issue here.

**Comment by petrochenkov** [comment]
## Stabilization report

All the features below target Rust 1.39 (November 7th 2019).
Tests can be found in test files touched by this PR.

### Fn-like macros and attribute macros in extern blocks

Macro attributes are now supported on items in `extern` blocks, the behavior is the same as for attributes on free/trait/impl items available on stable.

The feature name is `#![feature(macros_in_extern)]`.
The feature was implemented in April 2018 in https://github.com/rust-lang/rust/pull/49350 to fix issue https://github.com/rust-lang/rust/issues/48747.
It didn't go through an RFC because the feature seemed to be a trivial extension consistent with already supported macros in item and item-like positions.

### Attribute macros on inline modules

Macro attributes are now supported on inline module items (`mod m { ... }`), the behavior is the same as for attributes on any other items available on stable.

The feature is a part of the franken-feature `#![feature(proc_macro_hygiene)]`.
The feature was implemented together with all other proc macro attributes on items somewhere in 2017-2018.
Its current instability is a collateral damage from attributes on *out-of-line* modules, which actually have unresolved questions (like whether the `mod` file should be loaded before expanding the macro or after, and what should proc macros do if they need to load files themselves).

### Fn-like procedural macros in type positions

Proc macros can now be used in type positions (`type A = foo!();`), the behavior is the same as for `macro_rules` macros in the same position, which are available on stable.
The difference is that proc macros on stable use call-site hygiene, while `macro_rules` use mixed call-site/def-site hygiene, but that difference only affects local variables, labels and `$crate` so it doesn't apply to types (or at least applies to types no more than to item-position macros, which are available on stable).

The feature is also a part of the polyphyletic `#![feature(proc_macro_hygiene)]`.
The feature was implemented together with all other proc macros somewhere in 2017-2018.
It wasn't stabilized yet because it somehow got lost during the Macro 1.2 stabilization wave, and nobody actively pushed it after that.

**Comment by petrochenkov** [comment]
@Centril 
>it would be good to elaborate re. hygiene, how this may e.g. interact with expressions in type contexts, and give a recap (because issues re. hygiene isn't the best known subject...) of the issues wit proc macros expanding to expressions and why it isn't an issue here.

Proc macros in expressions and patterns have mostly philosophical issues - they can 1) consume local variables from their environment code and 2) can produce local variables consumable from their environment code.
Given that proc macros on stable can only use "unhygienic" `Span::call_site` hygiene, and unhygienic local variables are bad, proc macros in expression and patterns are prohibited in general.
Those philosophical issues do not apply to types, they can neither produce nor consume local variables.

The technical issue with expressions and patterns is that local variables stress the hygiene algorithm more and are more likely to hit some corner cases that may certainly exist because the hygiene algorithm does a few questionable things and is not fully understood.
That said, you still can work that feature gate around with stable item macros (https://crates.io/crates/proc-macro-hack), but some hope exists that proc-macro-hack is still a more limited exposure than arbitrary stable expression/pattern macros would be.

**Comment by petrochenkov** [comment]
>As such, I would like us to consider these separately in 3 different stabilization PRs with associated reports & FCPs.

The features are pretty trivial (that's why they were selected for immediate stabilization in the first place), I can split the PR in the lang team wants, but that would mostly create extra work for everyone involved, IMO.

**Comment by Centril** [comment]
> The features are pretty trivial (that's why they were selected for immediate stabilization in the first place), I can split the PR if the lang team wants, but that would mostly create extra work for everyone involved, IMO.

Fair enough; the types bit is the only bit I currently have worries about (see below).

> Those philosophical issues do not apply to types, they can neither produce nor consume local variables.

How does this affect interactions with VLAs in the style of https://github.com/rust-lang/rust/issues/48055 or run-time-value-dependent typing more generally? For example, `let x: type_mac!() = ...;` looks like it could "consume" a local `let n` if unhygienic and producing `[u8; dyn n]` or some such?

**Comment by petrochenkov** [comment]
@Centril 
>How does this affect interactions with VLAs in the style of https://github.com/rust-lang/rust/issues/48055 or run-time-value-dependent typing more generally? For example, `let x: type_mac!() = ...;` looks like it could "consume" a local `let n` if unhygienic and producing `[u8; dyn n]` or some such?

To clarify, `Span::call_site` hygiene generally works for local variables (I haven't seen bug reports for it at least), and already available on stable after jumping through a couple of hoops, e.g.
```rust
#[proc_macro]
fn item_macro(input) {
    "fn item() {
        ${input}
        let y = x; // OK    
    }"
}

item_macro!(let x = 10);
```
As I understand, the nrc's desire was to limit it (https://github.com/rust-lang/rust/pull/52121#issuecomment-422311424) rather than prohibit completely.
So further minor leaks like `[u8; dyn n]` should be ok.

**Comment by Centril** [comment]
First, sorry about the delay here but I think I need to understand better...

> As I understand, the nrc's desire was to limit it ([#52121 (comment)](https://github.com/rust-lang/rust/pull/52121#issuecomment-422311424)) rather than prohibit completely.

cc @nrc, @pnkfelix (maybe back from vacation soon?), @samth, and @rust-lang/lang 

>  So further minor leaks like `[u8; dyn n]` should be ok.

So I feel a distinct lack of understanding re. what minor constitutes here. Would e.g. `[u8; dyn { n = 3; n }]` change anything?

**Comment by nrc** [comment]
> As I understand, the nrc's desire was to limit it (#52121 (comment)) rather than prohibit completely.
> So further minor leaks like [u8; dyn n] should be ok.

To clarify, I don't have a specific objection, just a general level of uncertainty about our macro hygiene in general. I'd prefer to not stabilise anything until we resolve that uncertainty, but that might have happened since I was last involved (or might be unrealistic to expect).

**Comment by petrochenkov** [comment]
>So I feel a distinct lack of understanding re. what minor constitutes here. Would e.g. `[u8; dyn { n = 3; n }]` change anything?

Minor as in amount of code it's used in, so it less likely to hit corner cases statistically.
(`n = 3` doesn't change anything.)
Anyway, we can feature gate `dyn n` in expansions specifically, if it's ever implemented.

**Comment by eddyb** [comment]
`[T; dyn expr]` in types isn't at thing that I've ever seen proposed.

The VLA(-like) proposal was for `[expr; dyn expr]` evaluating to a `[T]` value.

Runtime expressions in types amount to dependent typing, which we can all agree is far away, if ever, and for the time being these proc macros don't really have issues like that.

<hr/>

That said, the interesting case IMO is this:
```rust
macro_rules! foo {
    ($x:ident, $e:expr) => ([(); {
        let $x = 0;
        let x = 1;
        $e
    }])
}

macro_rules! bar {
    ($x:ident, $e:expr) => ([(); {
        let $x = 0;
        let $x = 1;
        $e
    }])
}

const _: foo!(x, x + 1) = [()];
const _: bar!(x, x + 1) = [(), ()];
```
If you were to implement either `foo` or `bar` as proc macros, would they both behave like `bar`?

*However*, even if that is the case, since the expression using `x` must be passed into the macro for the distinction to matter, the proc macro could very well "scavenge" for the right identifier in the input to produce the unhygienic behavior anyway.

So it does help that types are "closed expressions", wrt local bindings in scope.

**Comment by petrochenkov** [comment]
I've just noticed that this PR stabilizes the next code as well:
```rust
#[my_attr]
mod m {
    mod n; // Out-of-line module, how should `my_attr` see it in its input?
}
```
So, we additionally need to feature gate out-of-line modules in attr/derive input.

This may affect stable in corner cases:
```rust
#[my_attr]
fn m() {
    #[path = "zzz.rs"]
    mod n;
}
```

**Comment by Centril** [comment]
> I've just noticed that this PR stabilizes the next code as well:

Hmm; can we move the inline module bits to a different PR then?

**Comment by Robbepop** [comment]
I currently see problems stabilizing custom attributes on inline modules due to https://github.com/rust-lang/rust/issues/43081.
The effect of this Rust compiler bug is that instead of receiving proper errors in your custom attributes that point to concrete spans of the source code you receive errors that completely lost all span information which is kind of bad for user experience.

**Comment by petrochenkov** [comment]
@Robbepop 
>I currently see problems stabilizing custom attributes on inline modules due to #43081.

#43081 affects all proc macros, why should it block stabilization of module macros specifically?

@Centril 
>can we move the inline module bits to a different PR then?

Done in https://github.com/rust-lang/rust/pull/64273.

**Comment by Robbepop** [comment]
> I currently see problems stabilizing custom attributes on inline modules due to #43081.

I might be confused about this and maybe I am missing details but I do not have any span information problems with any kind of proc. macros that are not applied to inline-modules. But for the inline-module macro all span information is missing. So if you are right and my examples do generally not apply I am fine with stabilization.

**Comment by Centril** [comment]
We discussed this proposal on this week's language team meeting.

@nrc attended and elaborated on their concerns in https://github.com/rust-lang/rust/pull/63931#issuecomment-527299574. To summarize, they felt uncertain about our hygiene story and many in the language team concurred that we don't have the best understanding of hygiene. In the best of worlds, we would perhaps want a PhD thesis around hygiene under the supervision of @samth to gain deep understanding. Realistically however, that might not materialize for a long time and would take even more time to complete.

We did however note that proc-macro-hack exists and so we don't expect that we would make the situation worse by stabilizing. On the upside, we were also able to transition to Rust 2018 and hygien-ize things. We noted that where problems might arise, it would be in the edge cases. I also noted on the meeting that some work has also been done by @petrochenkov and @matthewjasper since then so the situation should have improved a bit at least.

In summary, despite the aforementioned risks, we believe we are ready to propose a final comment period at this point.

@rfcbot merge

**Comment by petrochenkov** [comment]
@Robbepop 
>I might be confused about this and maybe I am missing details but I do not have any span information problems with any kind of proc. macros that are not applied to inline-modules.

Could you give some minimized reproduction of the span issues you are encountering?
Perhaps they are related to e.g. inner attributes or `cfg`s, that are just more likely to be used with modules, or something like that.

**Comment by Robbepop** [comment]
> @Robbepop
> 
> > I might be confused about this and maybe I am missing details but I do not have any span information problems with any kind of proc. macros that are not applied to inline-modules.
> 
> Could you give some minimized reproduction of the span issues you are encountering?
> Perhaps they are related to e.g. inner attributes or `cfg`s, that are just more likely to be used with modules, or something like that.

This is what I could come up with quickly:
https://github.com/Robbepop/missing_spans
Go to the `tests/ui` where you can find tests of this minimal proc macro and the expected output.
Note that for the test that uses a random inner attribute in the module body the span information is completely missing.

**Comment by petrochenkov** [comment]
@bors r=Centril
(This was actually r+'d in https://github.com/rust-lang/rust/pull/63931#issuecomment-526795276.)
